test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs

```diff
@@ -0,0 +1,336 @@
+    } deriving Show
+
+
+coinSelectionUnitTest
+    :: UnitTestInput
+    -> Expectation
+coinSelectionUnitTest (UnitTestInput n utxoCoins txOutsCoins expected) = do
+    (utxo,txOuts) <- setup
+
+    result <- runExceptT $ largestFirst
+              (defaultCoinSelectionOptions n)
+              utxo
+              txOuts
+
+    case expected of
+        Left err ->
+            result `shouldBe` (Left err)
+        Right expectedCoinsSel ->
+            (right (\(CoinSelection inps _  _) -> map (getCoin . coin . snd) inps) result)
+            `shouldBe`
+            (return expectedCoinsSel)
+    where
+        setup :: IO (UTxO, NonEmpty TxOut)
+        setup = do
+            txUtxoIns <- generate $ vectorOf (L.length utxoCoins) arbitrary
+            txUtxoOutsAddr <- generate $ vectorOf (L.length utxoCoins) arbitrary
+            let utxo = UTxO $ Map.fromList $ L.zip txUtxoIns
+                       $ L.zipWith TxOut txUtxoOutsAddr
+                       $ map Coin utxoCoins
+            txOutsAddr <- generate $ vectorOf (L.length txOutsCoins) arbitrary
+            let txOuts = NE.zipWith TxOut (NE.fromList txOutsAddr)
+                         $ NE.map Coin txOutsCoins
+            pure (utxo, txOuts)
+
+
+propDeterministic
+    :: CoveringCase
+    -> Property
+propDeterministic (CoveringCase (utxo, txOuts)) =
+    (condCoinsCovering txOuts utxo) ==> monadicIO $ liftIO $ do
+    resultOne <- runExceptT $ largestFirst
+                 (defaultCoinSelectionOptions 100)
+                 utxo
+                 txOuts
+    resultTwo <- runExceptT $ largestFirst
+                 (defaultCoinSelectionOptions 100)
+                 utxo
+                 txOuts
+
+    resultOne `shouldBe` resultTwo
+
+propAtLeast
+    :: CoveringCase
+    -> Property
+propAtLeast (CoveringCase (utxo, txOuts)) =
+    (condCoinsCovering txOuts utxo) ==> monadicIO $ liftIO $ do
+    result <- runExceptT $ largestFirst
+              (defaultCoinSelectionOptions 100)
+              utxo
+              txOuts
+
+    L.length
+        (concatMap (\ (CoinSelection inps _ _) -> inps) $ rights [result])
+        `shouldSatisfy`
+        (>= NE.length txOuts)
+
+propInputDecreasingOrder
+    :: CoveringCase
+    -> Property
+propInputDecreasingOrder (CoveringCase (utxo, txOuts)) =
+    (condCoinsCovering txOuts utxo) ==> monadicIO $ liftIO $ do
+    result <- runExceptT $ largestFirst
+              (defaultCoinSelectionOptions 100)
+              utxo
+              txOuts
+    let utxoEntriesInCoinSelection =
+            concatMap (\ (CoinSelection inps _ _) -> inps)
+            $ rights [result]
+    let utxo' =
+            Map.toList (getUTxO utxo) L.\\ utxoEntriesInCoinSelection
+    let getExtremumValue f = f . map (getCoin . coin . snd)
+
+    if L.null utxo' then utxo' `shouldBe` utxo' else
+        (getExtremumValue L.minimum utxoEntriesInCoinSelection)
+        `shouldSatisfy` (>= (getExtremumValue L.maximum utxo'))
```

Comment:
So, this checks that, the minimal selected value is greater than the maximum value from the remaining utxo. Sounds about right!

Reply:
used `unless`